Omni-directional CUAS jammer

b-mayr-1984 opened the pull request:
When merged adds an omni-directional CUAS jammer.

It is based on the previously named TFAR jammer (which I renammed to _voice communication jammer_).
When placing a jammer as Zeus or in 3den the `capabilities` of the jammer can be defined (defaults to being the same as the old TFAR jammer).
![image](https://github.com/Crowdedlight/Crows-Electronic-Warfare/assets/76476468/e77cda54-25ef-494f-bb89-02b67ee45ef3)

### Camera effect
When approaching the effective `radius` of an active `DroneJammer` the controling player's camera will be incrisingly distorted (see screenshots).

![image](https://github.com/gruppe-adler/Shoot_and_Scoot.Tanoa/assets/76476468/11695009-9065-4386-9947-e441ab2c3353)

![image](https://github.com/gruppe-adler/Shoot_and_Scoot.Tanoa/assets/76476468/88962170-4999-4e16-b6f3-161f1b7bd54c)

When the drone is within the effective `radius` the player get's disconnected (see screenshot).
![image](https://cdn.discordapp.com/attachments/402412803153199104/1202531922098786335/image.png?ex=65cdcc20&is=65bb5720&hm=ac8674aebedf7e16bd3f35fe5b06f5a78eb992687304a4c05b87abc1e2c2013b&)

### Effect on autonomous drones
Non-player controlled autonomous drones get their own server side handling via `fnc_jammerServerLoop.sqf`, which is quite in line with how jamming works with the Spectrum Device.

### Spectral behaviour
Drone jammers have a very distinct spectral behaviour in the Spectrum Device.
This shall allow players to locate those jammers (and maybe take them out).
In the code this is implemented as a new `beacon` type called `"sweep"`.
![image](https://github.com/Crowdedlight/Crows-Electronic-Warfare/assets/76476468/324ce56b-e2cf-4196-a37a-f0460b16c31d)

### Misc
- I added some minor fixes to existing code
- The `radius` marker for Zeus I made yellow in case the jammer is turned on (black otherwise; just as it were).
- `dev_mini.Enoch` has been expanded to test the new features

### Notes to maintainer
I was very careful not to affact the existing TFAR jammer functionality. But I did not test it in real multiplayer scenario with multiple human players yet. Same goes for the drone jammer which I only tested on my machine with me as a single client.
- Let me know if I shall join any of your test sessions.
- I will let you know when I did such tests. 

(But since this is such a large PR I think it has progressed enough to share what has been done and maybe discuss the changes and how they fit into `crowsEW`.)

Crowdedlight commented:
Looks great! I should be able to go through the PR this weekend! Especially love the detail of the sweeping jammer effect and the visual player effect on camera. 

If the player tries to reconnect to the drone after being kicked out, will it still attempt to connect but then kick them the player out again?

And are you including the "fail-safe" modes for drone-signals you suggested in this PR? Or what is the drone behaviour when it gets too close to the jammer? 
Will the AI and player drones be disabled and stay where they are, until jammer is moved out of distance, destroyed, or drone runs out of fuel?

Can player-controlled drones also be jammed by the spectrum device and give them the effect of visual degradation as long as the jammer is active, or is that behaviour only supported in the new omni-directional jammer?

-------------

btw, as you might have noticed I added a new wiki for this mod based on ``mdbook``. I have moved over the old wiki stuff and added a bit more. Please include some documentation on how the new jammer works, or is used. Essentially the description and images you have in this PR would do nicely ;-) 

You can see the current version of the new wiki here: https://crowdedlight.github.io/Crows-Electronic-Warfare/

You can add new pages in folders or new folders if necessary. Then add them as pages to the book by editing the SUMMERY.md file.

If you want to test it locally, you just need to install "mdbooks" with cargo install mdbook and then in the root of the repo just go mdbook serve and you can see your current state in localhost browser.
(Requires you got Rust/Cargo installed)

b-mayr-1984 commented:
### player reconnecting to jammed drone
> If the player tries to reconnect to the drone after being kicked out, will it still attempt to connect but then kick them the player out again?

Exactly. One can always re-connect to any jammed drone but is then disconnected immediately.
(Happens in `fn_jammerPlayerLocal.sqf` which runs every 0.5s)

### fail-safe behaviour
> And are you including the "fail-safe" modes for drone-signals you suggested in this PR? 

I assume you are talking about our discussion in #52. Is that correct?

>Or what is the drone behaviour when it gets too close to the jammer? Will the AI and player drones be disabled and stay where they are, until jammer is moved out of distance, destroyed, or drone runs out of fuel?

This PR does not change what I implemented earlier bd608cad490a97f13465fcf2fb7686ac5310b57c.
Drones will get a disabled AI and eventually land if being jammed for long enough. Once landed they just stay there. 
If the jammer moves out of range it can even be re-claimed via the UAV terminal and flown away.

### jamming via spectrum device
> Can player-controlled drones also be jammed by the spectrum device and give them the effect of visual degradation as long as the jammer is active, or is that behaviour only supported in the new omni-directional jammer?

I thought about this for a while. But jamming the RC uplink and the video downlink are 2 separate jamming tasks. 
The omni-directional jammer is meant as the big gun that does multi-band (or multi-application) jamming. 
I did not extend the current functionality of the spectrum device for the following reasons.
In order to degrade the video downlink the spectrum device would need to be pointed to the drone pilot, not the drone (if a somewhat realistic implementation is the goal). But so far the drone pilots location is not knowable for the spectrum device operator. It could be added that drone pilots are also `beacons`, so that their location or at least heading could be deduced, but for now I left this out.

But I think what I am currently missing is a forced disconnect from the drone of any controling player when the spectrum device is jamming the drone. Because right now the AI manipulation should be totally ignored by player controlled drones, which makes no sense. 
Thanks for the question that uncovered this gap. 🙂 

### Wiki
> btw, as you might have noticed I added a new wiki for this mod based on `mdbook`. I have moved over the old wiki stuff and added a bit more. Please include some documentation on how the new jammer works, or is used. Essentially the description and images you have in this PR would do nicely ;-)
> 
> You can see the current version of the new wiki here: https://crowdedlight.github.io/Crows-Electronic-Warfare/
> 
> You can add new pages in folders or new folders if necessary. Then add them as pages to the book by editing the SUMMERY.md file.
> 
> If you want to test it locally, you just need to install "mdbooks" with cargo install mdbook and then in the root of the repo just go mdbook serve and you can see your current state in localhost browser. (Requires you got Rust/Cargo installed)

Sure, I can give it a look. 🙂

b-mayr-1984 commented:
> But I think what I am currently missing is a forced disconnect from the drone of any controling player when the spectrum device is jamming the drone. Because right now the AI manipulation should be totally ignored by player controlled drones, which makes no sense. 

This has been implemented in ef8b62cae2286d39e602d4eb165bf689d4c8bd94.

Crowdedlight commented:
Reminder to self:

- ToggleJamming - JamLoop: I should probably rewrite this loop to be PFH to get out of scheduled space with current ``spawn``. Would make it easier to check if any jam-loops are running. 😅
- might make sense to move all the jamming out of ``main`` pbo and into a ``jamming`` pbo.

Crowdedlight commented:
> Exactly. One can always re-connect to any jammed drone but is then disconnected immediately. (Happens in `fn_jammerPlayerLocal.sqf` which runs every 0.5s)

Makes sense. 👍 

> This PR does not change what I implemented earlier [bd608ca](https://github.com/Crowdedlight/Crows-Electronic-Warfare/commit/bd608cad490a97f13465fcf2fb7686ac5310b57c). Drones will get a disabled AI and eventually land if being jammed for long enough. Once landed they just stay there. If the jammer moves out of range it can even be re-claimed via the UAV terminal and flown away.

Thanks for the clarification! 

> > Can player-controlled drones also be jammed by the spectrum device and give them the effect of visual degradation as long as the jammer is active, or is that behaviour only supported in the new omni-directional jammer?
> 
> I thought about this for a while. But jamming the RC uplink and the video downlink are 2 separate jamming tasks. The omni-directional jammer is meant as the big gun that does multi-band (or multi-application) jamming. I did not extend the current functionality of the spectrum device for the following reasons. In order to degrade the video downlink the spectrum device would need to be pointed to the drone pilot, not the drone (if a somewhat realistic implementation is the goal). But so far the drone pilots location is not knowable for the spectrum device operator. It could be added that drone pilots are also `beacons`, so that their location or at least heading could be deduced, but for now I left this out.

Yeah, that makes sense and is a fair interpretation. As mentioned in an earlier discussion I am sometimes a bit unsure about the balance of somewhat realistic implementation and gameplay loops. But I think this strikes a fair balance of just having the C2 link be jammable with the spectrum device.

> But I think what I am currently missing is a forced disconnect from the drone of any controling player when the spectrum device is jamming the drone. Because right now the AI manipulation should be totally ignored by player controlled drones, which makes no sense. Thanks for the question that uncovered this gap. 🙂

Glad my questions helped! :D 

> > btw, as you might have noticed I added a new wiki for this mod based on `mdbook`. I have moved over the old wiki stuff and added a bit more. Please include some documentation on how the new jammer works, or is used. Essentially the description and images you have in this PR would do nicely ;-)

> Sure, I can give it a look. 🙂

Would be much appreciated!


I added a few comments, very few code changes, mostly optional suggestions or ideas. Thanks for your big work in adding this to the mod. I quite like the look of it and I am sure players will like the additional features and effects you have added! :)

b-mayr-1984 commented on .gitignore:
```diff
@@ -13,5 +13,3 @@ hemtt
 *.dll
 .hemttout/*
-.vscode/sqfvm-lsp/
-target
```
Has nothing to do with my changes. 
I only changed `.gitignore` to enable the VSC debugger (see 2ec2656dccaa8a27195ae2a9864816c8db624e70).

Crowdedlight replied:
That is super odd. I must have made a mistake somewhere in previous push to develop somehow. I'll fix it after this is merged.

b-mayr-1984 commented on addons/main/functions/fnc_actionJamToggleListener.sqf:
```diff
@@ -19,4 +19,12 @@ private _jammer = GVAR(jamMap) get _netId;
+
+// show jammers in the spectrum
+if (_enabled) then {
+	[QEGVAR(spectrum,addBeacon), [_jamObj, 433, 300, "sweep"]] call CBA_fnc_globalEventJIP;
```
I honestly don't know. 

I haven't tested anything with regards to the TFAR jammer. 
a) I don't know the exact functionality that it needs to have.
b) In order to test you need a bunch of people (which I never had available; nor a server were I could quickly change crowsEW).

I _assume_ it will behave as you wrote. 
In addition I made this comment in `fnc_spectrumTrackingLocal.sqf`.
```sqf
/*******************************************************************/
// TODO implement different frequency band for VoiceCommJammers
// (those are is currently shown in the drone band)
/*******************************************************************/
```

Crowdedlight replied:
Fair enough. Will test it out on a server. As I read it a jammer would always show a sweep now if enabled. Although that is also fine. Would make sense to see some kind of either sweep or just high noise broadcast for radio jamming.

b-mayr-1984 commented on addons/main/functions/fnc_addJammer.sqf:
```diff
@@ -9,16 +9,16 @@ Return: none
 // add action 
-_unit addAction ["<t color=""#FFFF00"">Pull Out Wires", FUNC(actionJamToggle), [_netId], 7, true, true, "", format ["([%1] call %2)", str(_netId), FUNC(isJammerActive)], 6];
-_unit addAction ["<t color=""#FFFF00"">Duct Tape Wires Back In", FUNC(actionJamToggle), [_netId], 7, true, true, "", format ["!([%1] call %2)", str(_netId), FUNC(isJammerActive)], 6];
+_unit addAction ["<t color=""#FFFF00"">De-activate jammer", FUNC(actionJamToggle), [_netId], 7, true, true, "", format ["([%1] call %2)", str(_netId), FUNC(isJammerActive)], 6];
+_unit addAction ["<t color=""#FFFF00"">Activate jammer", FUNC(actionJamToggle), [_netId], 7, true, true, "", format ["!([%1] call %2)", str(_netId), FUNC(isJammerActive)], 6];
```
Exactly. When I saw this action on the Wiesel, that I originally intended to use as the jammer, I appeared totally out of place.
![image](https://static.wikia.nocookie.net/armedassault/images/2/2a/Arma3-render-nyxrecondigitalgreen.png/revision/latest?cb=20190816092657)

b-mayr-1984 commented on addons/main/functions/fnc_jammerPlayerLocal.sqf:
```diff
@@ -77,28 +77,65 @@ private _distRad = -1;
+		private _txInterference = 1;
+
+		// for now staying with linear degradation of signal. Might make it a tad better for players than the sudden commms -> no comms exponential could induce
+		private _rxInterference = _jamStrength - (_distPercent * _jamStrength) + 1; 	// recieving interference. above 1 to have any effect.
+		private _txInterference = 1 / _rxInterference; 									// transmitting interference, below 1 to have any effect.
```
Roger

b-mayr-1984 replied:
Done in 08f3f1c696c868ed6b0ad56a1b41292fda191d74

b-mayr-1984 commented on addons/main/functions/fnc_jammerPlayerLocal.sqf:
```diff
@@ -77,28 +77,65 @@ private _distRad = -1;
+if (!isNull _drone) then {
+	// sort drone jammers by distance to drone not distance to player (which only makes sense for VoiceCommJammers)
+	private _sortingCode = { _input0 distance (_x select 0) };	// sort by distance between drone and jammer
+	private _filterCode = { (_x select 3)  && ("DroneJammer" in (_x select 4)) };	// keep jammers that are enabled and have the "DroneJammer" capability
```
When I began working on crowsEW I wanted to keep it consistent with how you did it.
Then when I learned that `#` has a higher operator preference than `select` I started using it more because you safe on a lot of `()` which you don't need anymore.

Crowdedlight replied:
Fair! Yeah, I believe I wrote a fair bit of this code before I learned of ``#`` so my use is a bit of a mix. I don't have a great code-style standard applied to this mod as I learned stuff as I developed it, and didn't refactor old stuff specifically 😅

b-mayr-1984 replied:
> You could also consider doing lazy eval for the if-statement, to avoid an array `in` command if the jammer is not enabled anyway.
> 
> ```sqf
> private _filterCode = { (_x select 3)  && { ("DroneJammer" in (_x select 4)) } };
> ```

Done in 05e235a4b262bd49f46bb463dd854406dc089bfa

b-mayr-1984 commented on addons/spectrum/functions/fnc_spectrumDeviceMouseDown.sqf:
```diff
@@ -125,6 +125,15 @@ private _timeActive = 5;
 			breakOut "loopFreq";
 		};
+		case "sweep": {
+			// players can not interact with a drone jammer
+			// so just ignore this signal
```
Agreed. 
It is also really hard to hit the peak of the jammer when it sweeps by.

Having a nice [frequency chirp](https://en.wikipedia.org/wiki/Chirp) sound would be plausible but would require some re-writing to make it viable for the players to trigger it.

Crowdedlight replied:
yeah, that makes sense. Would be hard to hit with the existing approach to triggering the sound.

b-mayr-1984 commented on addons/spectrum/functions/fnc_spectrumEnableSettingChanged.sqf:
```diff
@@ -29,7 +29,7 @@ if (_value) then {
 
 	// due to best practices we are gonna put the track loop in unscheduled space. 
-	GVAR(PFH_beaconPlayer) = [FUNC(spectrumTrackingLocal), 0.2] call CBA_fnc_addPerFrameHandler; 
+	GVAR(PFH_beaconPlayer) = [FUNC(spectrumTrackingLocal), 0.1] call CBA_fnc_addPerFrameHandler; 
```
I also tried it with `0.1` but on my machine I did not get a significantly nicer sweep display. 
`0.2` looked almost the same and was easier on my FPS, so I stuck with it.

Does your setup look nicer with `0.1`?

b-mayr-1984 replied:
Oops... my bad. 😊 
I thought you suggested 0.1. 

I changed the repetition interval because the original interval makes the sweep look ugly (to few peaks shown before looping back or to slow a frequency increment when insisting on more peaks).

`0.1` was the best tradeoff between beauty and FPS that I could find.

This PFH parameter and `_sweepIncrementSpeed` in `fnc_getNextSweepFreq.sqf` can be somewhat balanced against each other if you want to tinker with those.

b-mayr-1984 commented on addons/spectrum/functions/fnc_spectrumTrackingLocal.sqf:
```diff
@@ -56,15 +56,24 @@ private _sigsArray = [];
-    // if jammer is equipped, only show signals that is type drone 
-    if (GVAR(spectrumRangeAntenna) == 3 && _type != "drone") then { continue; };
+    // if jammer is equipped, only show signals that is type drone or sweep
+    if (GVAR(spectrumRangeAntenna) == 3 && !(_type in ["drone", "sweep"])) then { continue; };
```
Roger

b-mayr-1984 replied:
Done in 05e235a4b262bd49f46bb463dd854406dc089bfa

b-mayr-1984 commented on addons/spectrum/functions/fnc_spectrumTrackingLocal.sqf:
```diff
@@ -56,15 +56,24 @@ private _sigsArray = [];
+    if (_type == "sweep") then {
+        _frequency = [433, 7, 5, _forEachIndex] call FUNC(getNextSweepFreq);    // overides the original _frequency value
+		/*******************************************************************/
+		// TODO implement different frequency band for VoiceCommJammers
```
I **don't** plan to do this in the scope of this PR.

I simply lack a baseline for the behaviour of the TFAR jammer. 
I don't want to tinker around in a feature that I did not take the time to fully understand (will only result in bad code quality).

Crowdedlight replied:
Completely fair. Was mainly curious and wanted to know if more was to come here, or you are fine with merging this PR as it is :)

b-mayr-1984 commented on addons/spectrum/functions/fnc_toggleJammingOnUnit.sqf:
```diff
@@ -86,6 +88,15 @@ if (_enableJam) then {
+			private _droneUsers = (UAVControl _unit) select { !(_x isEqualType "String") };	 // leave only player objects
+			{
+				_x connectTerminalToUAV objNull; // disconnect player from drone
+				if (hasInterface) then {
```
> This event is only raised on the server, so `hasInterface` would always return false, as you are checking the server, not the player.

Oh it is even worse than that. 😅 
When being a Spectrum Device jammer yourself and playing as both client and server you actually see this message every time you jam any drone (no matter if you are controlling it), because `hasInterface` is then actually `true`.
I made it a little less bad by d2f1d6782ac250eee34064eb505bade94cc08ffe.

> You could use Zeus Enhanceds hint event to send the hint as a `targetEvent` on the player. probably the easiest solution and avoids remoteExec. https://github.com/zen-mod/ZEN/blob/13fb2466794d64ac6cfac3e26bda36d9efcbd152/addons/common/XEH_postInit.sqf#L31C1-L34C33

I have to look into that. 🙂 
Dealing with such locality issues is still giving me a hard time if I don't just align with what you already did.

Crowdedlight replied:
You can do:
```sqf
// _message is used by ZEN as: hint _message
["zen_common_hint", [_message], _targetPlayer] call CBA_fnc_targetEvent;
```
And it will send an hint only to the player given with ``_targetPlayer``, so passing in the players object there would do the trick. 

``player`` will always return your "local" player in MP. So using ``player`` on a client, will return that clients player. So if you do ``player in _droneUsers`` on the server, as you are doing in  https://github.com/Crowdedlight/Crows-Electronic-Warfare/commit/d2f1d6782ac250eee34064eb505bade94cc08ffe then you are checking if the "servers" player, is present in that array and then showing hint to the server. Which won't work as on dedicated servers, server has no interface, and I actually don't know what ``player`` will return if on a dedicated server 😅 

I would suggest doing:
```sqf
private _droneUsers = (UAVControl _unit) select { !(_x isEqualType "String") };	 // leave only player objects
{
	_x connectTerminalToUAV objNull; // disconnect player from drone
	["zen_common_hint", ["Drone is jammed<br/><t color='#ff0000'>Connection lost</t>"], _x] call CBA_fnc_targetEvent; // hint to player why they got disconnected
} forEach _droneUsers;
```

As this sends an single event to the eventhandler with id "zen_common_hint" on the targets client, so on the players client. And there the Eventhandler I linked from ZEN then executes:
```sqf
    params ["_message"];
    hint _message;
```

b-mayr-1984 replied:
Fixed in 8f9e256cb023d4400cea79eab064a200b4719b7c
(I just needed to add `parseText` to your code to make it work.)

b-mayr-1984 commented on addons/zeus/functions/fnc_addJammerZeus.sqf:
```diff
@@ -36,18 +39,25 @@ private _onConfirm =
 	};
 
+	private _capabilities = [];	// what types of signals can this jammer counteract?
+	if (_isVoiceCommsJammer) then { _capabilities pushBack "VoiceCommsJammer" };
+	if (_isDroneJammer) then { _capabilities pushBack "DroneJammer" };
```
Hm don't know. 
In C/C++ I would do it using enums.
In SQF I couldn't find those.

I chose this approach because it makes the code easily readable.

If what you are suggesting is a good idea or not I can not tell. 
I simply don't feel qualified enough with SQF and it's intermingling with C++ (and how the compiler works) to judge.

Crowdedlight replied:
Fair enough. I'll suggest going with ``#defines`` as you are right that enums outright doesn't exist in sqf. So would make sense for define to avoid typo mistakes in future. I'll see if I can do ``suggest changes`` on this PR to directly change for that, so you just have to click accept. I'll do that when I get home from work tomorrow.

Crowdedlight commented on addons/main/functions/fnc_jammerPlayerLocal.sqf:
```diff
@@ -77,28 +77,65 @@ private _distRad = -1;
+private _drone = getConnectedUAV player;
+if (!isNull _drone) then {
+	// sort drone jammers by distance to drone not distance to player (which only makes sense for VoiceCommJammers)
+	private _sortingCode = { _input0 distance (_x select 0) };	// sort by distance between drone and jammer
```
ah, I read that page, but missed that small part. My bad.

Crowdedlight commented:
> > * might make sense to move all the jamming out of `main` pbo and into a `jamming` pbo.
> 
> That might be a beginner question: What are the advantages of separating the functionality into different PBOs?
> 
> I noticed that I sometimes struggled when the jammer functionality was spread across `main` and `spectrum`. Not only did I need to `#define DISABLE_COMPILE_CACHE` twice but also the macro handling like `GVAR` vs. `EGVAR` etc. was confusing at times.

For this case mainly just structural in being easier to find and figure out where behaviour is defined and to not have 60+ sqf files in the same folder for all the different features. 
Technically you could delete individual PBOs from the mod, and then it just wouldn't load that part of the mod. Like deleting the "medicals" pbos from ace. But this mod is a tad too integrated with itself to assume that would work 😅 

Its more that afterwards I learned that "main" is often more used for common functions, variables or stuff shared across pbos/modules. 

I am not going to change it now, it was more if we start having fairly more features added regarding jamming, it would make sense to move it to a "jamming section" organisational. 

Fair point about it being more confusing when split up, I am obviously biased from making most of it, to just know and assume where stuff is. 😅 

I should probably figure out if I can get it working setting a single ``#define DISABLE_COMPILE_CACHE`` in "main" and then it works for all of them. Technically that should be the case as I believe each ``script_component.hpp`` should ``#include`` back to the one in ``main``. But have not tested it. 

``GVAR`` vs ``EGVAR`` also took some time for me to get used to. But kinda like it now, as I find it better using that then remembering to use "external gvar" or "quoted gvar" with ``EGVAR`` and ``QGVAR``.

Crowdedlight commented on addons/editormodules/functions/fnc_addJammer.sqf:
```diff
@@ -42,14 +42,22 @@ if (_activated) then {
+	private _isDroneJammer = _logic getVariable ["IsDroneJammer",0];
+
+	private _capabilities = [];	// what types of signals can this jammer counteract?
+	if (_isVoiceCommsJammer) then { _capabilities pushBack "VoiceCommsJammer" };
+	if (_isDroneJammer) then { _capabilities pushBack "DroneJammer" };
```
Please add this to PR. (I can't do it from here, as it doesn't have changes)

Add the following to ``script_macros.hpp`` file in ``main`` (``/addons/main/script_macros.hpp``)

```hpp
// Defines for Jamming capabilities for the Radio and Drone jammer. By having it here its automatically included and available to all submodules  

#define JAM_CAPABILITY_RADIO QUOTE(VoiceCommsJammer)
#define JAM_CAPABILITY_DRONE QUOTE(DroneJammer)
```

b-mayr-1984 replied:
Done in 30dbcd96b11a1e97b25c46e8174f6bb1930b8418

Crowdedlight commented on addons/main/functions/fnc_jammerPlayerLocal.sqf:
```diff
@@ -77,28 +77,65 @@ private _distRad = -1;
 	[player] call FUNC(resetTfarIfDegraded);
+} else {
+	// check for jammer capabilities and counteract signals accordingly
+	if ( "VoiceCommsJammer" in (_nearestJammer select 4) ) then {
```
```suggestion
	if ( JAM_CAPABILITY_RADIO in (_nearestJammer select 4) ) then {
```

Crowdedlight commented on addons/zeus/functions/fnc_addJammerZeus.sqf:
```diff
@@ -36,18 +39,25 @@ private _onConfirm =
 	};
 
+	private _capabilities = [];	// what types of signals can this jammer counteract?
+	if (_isVoiceCommsJammer) then { _capabilities pushBack "VoiceCommsJammer" };
+	if (_isDroneJammer) then { _capabilities pushBack "DroneJammer" };
```
```suggestion
	if (_isVoiceCommsJammer) then { _capabilities pushBack JAM_CAPABILITY_RADIO };
	if (_isDroneJammer) then { _capabilities pushBack JAM_CAPABILITY_DRONE };
```

Crowdedlight commented on addons/main/functions/fnc_jammerPlayerLocal.sqf:
```diff
@@ -77,28 +77,65 @@ private _distRad = -1;
+if (!isNull _drone) then {
+	// sort drone jammers by distance to drone not distance to player (which only makes sense for VoiceCommJammers)
+	private _sortingCode = { _input0 distance _x#0 };	// sort by distance between drone and jammer
+	private _filterCode = { _x#3  && { "DroneJammer" in _x#4 } };	// keep jammers that are enabled and have the "DroneJammer" capability
```
```suggestion
	private _filterCode = { _x#3  && { JAM_CAPABILITY_DRONE in _x#4 } };	// keep jammers that are enabled and have the "DroneJammer" capability
```

b-mayr-1984 commented:
> btw, as you might have noticed I added a new wiki for this mod based on `mdbook`. I have moved over the old wiki stuff and added a bit more. Please include some documentation on how the new jammer works, or is used. Essentially the description and images you have in this PR would do nicely ;-)
> 
> You can see the current version of the new wiki here: https://crowdedlight.github.io/Crows-Electronic-Warfare/
> 
> You can add new pages in folders or new folders if necessary. Then add them as pages to the book by editing the SUMMERY.md file.
> 
> If you want to test it locally, you just need to install "mdbooks" with cargo install mdbook and then in the root of the repo just go mdbook serve and you can see your current state in localhost browser. (Requires you got Rust/Cargo installed)

I followed your hyperlink and then used the `Suggest an edit` button on the upper right corner to add my changes (see 
8e91a5ba162cf03e47cabce5f9d063f6e930b830). This however branched off from `main` again (sorry for the inconvenience 🙄)

b-mayr-1984 commented:
> Considering right now to move to remove develop, and just have "main". As the releases are done in tags anyway. Would make it a bit simpler in future. Not completely sure yet though 😅

I would very much appreciate that. 🙂 
A lot of tooling is simply expecting this as the standard workflow. And that is also what most developers are used to.

b-mayr-1984 commented:
We did a test with 3 people on Gruppe Adler infrastructure today. There is still a bunch of problems with PR. 
**Mostly JIP and locality issues.**

- [ ]  activation/deactivation state of omni jammer is not JIP safe
- [ ] jamming of autonomous drones with the omni jammer does not work at all with a dedicated server (probably due to locality issue with the drone)
- [ ] jamming of player controlled drones with the Spectrum Device does not work on dedicated server. No disconnect happens, even though the disconnect hint is displayed. (suspected locality issue)

Crowdedlight commented:
> > Considering right now to move to remove develop, and just have "main". As the releases are done in tags anyway. Would make it a bit simpler in future. Not completely sure yet though 😅
> 
> I would very much appreciate that. 🙂 A lot of tooling is simply expecting this as the standard workflow. And that is also what most developers are used to.

I have no arguments to not do it like that 😅. So I will do that in future. When this PR is merged I'll do a merge into main, and stage everything from main in the future. 

In regards to the locality and JIP issues. I'll have a closer look at the communication flow. not JIP safe, do you mean that players joining as JIP don't get affected at all? Or how does this express itself? (I'll have a look to see if the event broadcast is using Server/global events, or GlobalJIP events)

When you say: "omni jamming of drones doesn't work at all on dedicated server", Is that then the AI part? 

Hmmm, odd, as when I checked the disconnect command for player controlled UAV it is global effect and global argument, so I would expect that to work no matter the locality.

Crowdedlight commented:
> Yes. That is why I wrote _autonomous drones_ .

Sorry, didn't catch that part. Thanks for clarifying THAT it was BOTH behaviour and no jammed text. 

> Same here 😕 I also saw a bunch of error messages in the server RPT that related to UAV objects that it couldn't find. Could be related to this issue.

Might be some special cases around the uav terminal and locality.

Crowdedlight commented on addons/main/functions/fnc_addJammer.sqf:
```diff
@@ -44,8 +47,13 @@ if (typeof _unit == "Land_DataTerminal_01_F") then {
+
+if (_enabled) then {
+	// add jammer as a signal beacon (so that it can be tracked down with the Spectrum Device)
+	[QEGVAR(spectrum,addBeacon), [_unit, 433, 300, "sweep"]] call CBA_fnc_globalEventJIP;
```
I think this can be part of the current issue. This ``addJammer`` function is called with a ``globalEventJIP``, which means every single client will execute what it have. This in turns means that if its enabled, the ``addBeacon`` gets executed for every single client connected to the game multiplied by the amount of players. 
So with 4 players online, it would be called ~16 times for the same jammer. 

It should work for the ``GVAR(jamMap)`` because that is a local variable that gets updated.

I need to have a closer look, but might be possible to add ``isServer`` to the if-statement, so signal is only added if enabled and from the server-instance. Then the addBeacon ``globalEventJIP`` gets sent to all other clients. 
Alternatively might be possible to just run the ``addBeacon`` function locally for everyone, and skip the event broadcast.

b-mayr-1984 replied:
The CBA guys recommend the following [here](https://github.com/CBATeam/CBA_A3/wiki/Custom-Events-System):
_For state-changes, which you would want synced on JIP, use the standard `publicVariable` and `addPublicVariableEventHandler` commands instead._

This sounds like it applies very well to our situation with the jammer state. I don't see a reason why we would need the [stacking feature](https://cbateam.github.io/CBA_A3/docs/files/events/fnc_globalEventJIP-sqf.html) of `globalEventJIP`. Noone needs the information how many times a jammer has been turned on and off before JIPing. They only need to know the current state.

Crowdedlight commented:
When the major restructuring work has been merged from https://github.com/b-mayr-1984/Crows-Electronic-Warfare/pull/1 Then this should be ready for merge. 

The patch has verified working code tested on dedicated server. 

The only think that could be nice to add in this PR before merge, would be updating the wiki with the changes from the new drone behaviour and options.

Crowdedlight commented:
@b-mayr-1984 Would you have time to update the wiki with the changes this PR makes? Like just your changes to the jammers and addition of omni-directional jammer. So there is some documentation on players can get jammed now etc.

b-mayr-1984 commented:
> @b-mayr-1984 Would you have time to update the wiki with the changes this PR makes? Like just your changes to the jammers and addition of omni-directional jammer. So there is some documentation on players can get jammed now etc.

You can find my Wiki changes here https://github.com/b-mayr-1984/Crows-Electronic-Warfare/commit/8e91a5ba162cf03e47cabce5f9d063f6e930b830
